src/main/java/uk/gov/justice/hmpps/prison/service/AdjudicationsService.java

```diff
@@ -264,4 +323,167 @@ private AdjudicationDetail transformToDto(final Adjudication adjudication) {
+
+        idsToRemove(requiredConnectedOffenderBookings, adjudication.getConnectedOffenderBookings(), OffenderBooking::getBookingId)
+            .forEach(offenderBookingId -> remove(remainingConnectedOffendersParties, AdjudicationParty::offenderBookingId, offenderBookingId));
+
+        return List.of(
+            remainingVictimsStaffParties,
+            remainingVictimsOffenderParties,
+            remainingConnectedOffendersParties
+            ).stream().flatMap(Collection::stream).toList();
+    }
+
+    private Optional<Staff> staffWithId(List<Staff> staff, Long id) {
+        return staff.stream().filter(s -> id.equals(s.getStaffId())).findFirst();
+    }
+
+    private Optional<OffenderBooking> offenderBookingWithId(List<OffenderBooking> offenderBookings, Long id) {
+        return offenderBookings.stream().filter(o -> id.equals(o.getBookingId())).findFirst();
+    }
+
+    private AdjudicationParty newConnectedOffenderAdjudicationParty(Adjudication adjudication, AtomicReference<Long> sequence, LocalDateTime currentDateTime, OffenderBooking offenderBooking) {
+        return newAdjudicationParty(adjudication, sequence, currentDateTime)
+            .incidentRole(Adjudication.INCIDENT_ROLE_OFFENDER)
+            .offenderBooking(offenderBooking).build();
+    }
+
+    private AdjudicationParty newVictimStaffAdjudicationParty(Adjudication adjudication, AtomicReference<Long> sequence, LocalDateTime currentDateTime, Staff victimStaff) {
+        return newAdjudicationParty(adjudication, sequence, currentDateTime)
+            .incidentRole(Adjudication.INCIDENT_ROLE_VICTIM)
+            .staff(victimStaff).build();
+    }
+
+    private AdjudicationParty newVictimOffenderAdjudicationParty(Adjudication adjudication, AtomicReference<Long> sequence, LocalDateTime currentDateTime, OffenderBooking offenderBooking) {
+        return newAdjudicationParty(adjudication, sequence, currentDateTime)
+            .incidentRole(Adjudication.INCIDENT_ROLE_VICTIM)
+            .offenderBooking(offenderBooking).build();
+    }
+
+    private AdjudicationParty.AdjudicationPartyBuilder newAdjudicationParty(Adjudication adjudication, AtomicReference<Long> sequence, LocalDateTime currentDateTime) {
+        sequence.set(sequence.get() + 1);
+        return AdjudicationParty.builder()
+            .id(new AdjudicationParty.PK(adjudication, sequence.get()))
+            .partyAddedDate(currentDateTime.toLocalDate());
+    }
+
+    private <T> Set<Long> idsToAdd(List<T> required, List<T> current, Function<T, Long> toId) {
+        return idsToAdd(required.stream().map(toId).toList(), current.stream().map(toId).toList());
+    }
+
+    private <T> Set<Long> idsToRemove(List<T> required, List<T> current, Function<T, Long> toId) {
+        return idsToRemove(required.stream().map(toId).toList(), current.stream().map(toId).toList());
+    }
+
+    private Set<Long> idsToAdd(List<Long> desired, List<Long> current) {
```

Comment:
Yeah, reason I didn't was because the old code could deal with multiple same values before I realised the restriction - and I didn't change it all to sets - but I digress. Yes sets should be what we use